File: LLD/Design-Patterns/TemplateMethod/code/template_method.py
"""
Template Method — CaffeineBeverage fixes the recipe sequence; Tea and Coffee fill in the varying steps.
Run: python template_method.py
"""

from abc import ABC, abstractmethod


class CaffeineBeverage(ABC):
    def prepare_recipe(self):
        """The template method — the fixed sequence of steps."""
        self.boil_water()
        self.brew()
        self.pour_in_cup()
        if self.wants_condiments():
            self.add_condiments()

    def boil_water(self):
        print("Boiling water")

    def pour_in_cup(self):
        print("Pouring into cup")

    @abstractmethod
    def brew(self):
        ...

    @abstractmethod
    def add_condiments(self):
        ...

    def wants_condiments(self):
        """Hook — subclasses may override; default is True."""
        return True


class Tea(CaffeineBeverage):
    def brew(self):
        print("Steeping the tea")

    def add_condiments(self):
        print("Adding lemon")


class Coffee(CaffeineBeverage):
    def brew(self):
        print("Dripping coffee through filter")

    def add_condiments(self):
        print("Adding sugar and milk")

    def wants_condiments(self):
        return False


if __name__ == "__main__":
    print("Preparing tea:")
    tea = Tea()
    tea.prepare_recipe()

    print("Preparing coffee (no condiments):")
    coffee = Coffee()
    coffee.prepare_recipe()

File: LLD/Design-Patterns/TemplateMethod/code/TemplateMethod.cpp
// Template Method — CaffeineBeverage fixes the recipe sequence; Tea and Coffee fill in the varying steps.
// Compile: g++ -std=c++14 TemplateMethod.cpp -o template_method
// Run:     ./template_method

#include <iostream>

class CaffeineBeverage {
public:
    // The template method — not virtual, so subclasses can't change the sequence.
    void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        if (wantsCondiments()) {
            addCondiments();
        }
    }

    virtual ~CaffeineBeverage() = default;

protected:
    void boilWater() {
        std::cout << "Boiling water" << std::endl;
    }

    void pourInCup() {
        std::cout << "Pouring into cup" << std::endl;
    }

    virtual void brew() = 0;
    virtual void addCondiments() = 0;

    // Hook — subclasses may override; default is true.
    virtual bool wantsCondiments() {
        return true;
    }
};

class Tea : public CaffeineBeverage {
protected:
    void brew() override {
        std::cout << "Steeping the tea" << std::endl;
    }

    void addCondiments() override {
        std::cout << "Adding lemon" << std::endl;
    }
};

class Coffee : public CaffeineBeverage {
protected:
    void brew() override {
        std::cout << "Dripping coffee through filter" << std::endl;
    }

    void addCondiments() override {
        std::cout << "Adding sugar and milk" << std::endl;
    }

    bool wantsCondiments() override {
        return false;
    }
};

int main() {
    std::cout << "Preparing tea:" << std::endl;
    Tea tea;
    tea.prepareRecipe();

    std::cout << "Preparing coffee (no condiments):" << std::endl;
    Coffee coffee;
    coffee.prepareRecipe();

    return 0;
}
